### ISSUE TITLE
does Criteria has support jdbc? 

### ISSUE THREAD

**Issue by zlfsimon** [immutables/immutables#1257]
Hi 
I want to use criteria/jdbc to connect to a mysql or postgresql database, and document said "Define criteria once and apply to different data-sources (Map, **_JDBC_**, Mongo, Elastic etc.)", but seems JDBC feature is not available now,  is it right?

**Comment by elucash** [comment]
Yep, seems to be, I stated once, but as I remember there's not much of a progress, there are tentative plans to return to it, but no guarantee when (

**Comment by zlfsimon** [comment]
oh， get it, thanks a lot. I think jdbc support is very important,  maybe we can't choose this framework even  it's very nice if lost supports of jdbc.  Anyway it's still a cool framework^^

**Comment by asereda-gs** [comment]
We do plan to support it, right now there is not enough time to implement it

**Comment by gtnicol** [comment]
I have a SQL backend implemented with query generation etc. Based on the discussion around Path it sounds like the plan was to use annotations to drive the table mapping generation?

**Comment by asereda-gs** [comment]
> Based on the discussion around Path it sounds like the plan was to use annotations to drive the table mapping generation?

Discussion around Path (#1256 ) is to find a way to define external models without java Reflection API. Ie use Criteria as a query translator layer.

If we have a generic model API (including type information) it can be based on reflection (annotations) or provided externally.

JDBC table mapping is related but not same.

**Comment by andonimichael** [comment]
Any update here? I'm also looking to use JDBC. And if you haven't started, I'd be willing to help contribute.

**Comment by gtnicol** [comment]
I do have a pretty complete implementation of SQL with support for MySQL and BigQuery

**Comment by elucash** [comment]
Any PRs or links to the separate repositories implementing backend module (for SQL) are very welcome. Thank you!

**Comment by gtnicol** [comment]
I'll check this code in over the next week. From memory I have BigQuery and MySQL dialects supported at this point.

**Comment by gtnicol** [comment]
How should I submit the SQL/JDBC criteria support? I can fork, or perhaps check into a feature branch? I'd need write permissions for the latter.

**Comment by elucash** [comment]
If you fork, do/apply changes to the fork (in any branch), then you can create PR (pull request) to merge it into our master, after review and any necessary changes we will merge it

**Comment by gtnicol** [comment]
PR is in https://github.com/immutables/immutables/pull/1394

**Comment by elucash** [comment]
thank you for PR, we will be reviewing

**Comment by gtnicol** [comment]
I'll do some testing on jdk8. I was thinking of doing a gremlin implementation... any thoughts on how best to approach graph models in criteria?

**Comment by elucash** [comment]
What I've learned over time is it's best to avoid any object-relational, object-graph, and any other barely compatible  models. What people usually need (with tools like Criteria) is a sort of statically typed queries and result-set mapping to objects (immutable data). I've looked only briefly at Gremlin, I don't know it it worth it to try to shoehorn graph querying into Criteria bases on (more simple) document model. If you use Gremlin java DSL, then you can somewhat reduce error-prone string attributes by using constants provided by (still experimental) `org.immutables:data` Datatypes generator, the same datatype metadata can be used to make reading of result set into objects easier. I just cannot say anything about just creating first-class Criteria implementation for graph querying, idk, to much to wrap my head around as of now.
Here's some features of the datatype metadata in action (in tests)
https://github.com/immutables/immutables/blob/master/data/test/org/immutables/data/DataTest.java

**Comment by gtnicol** [comment]
Makes sense. I tend to agree about ORMs and OGMs in general, hence the question. I was thinking of mapping an object to a node, but the edge modeling would get very weird .
